### PR TITLE
Laboratory work #3, Polina Kobyakova - 22FPL2

### PR DESCRIPTION
**Pull request by kobyakovapolina**

### CODE REVIEW

**Comment by kobyakovapolina** [comment]
I currently have some problems with pushing, but I've done several functions. Hoping to fix them.

**Review by artyomtugaryov — COMMENTED** [review]
Nothing to review. If you have problems with push - let's talk on the next seminar

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -159,6 +254,21 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+
+        text_new = []
+
+        for i in decoded_corpus:
+            text_new += i
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:263]: use append

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,28 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str):
+            return None
+
+        text = text.lower()
+        tokens = []
+
+        for word in text.split():
+            word = ''.join(letter for letter in word if letter.isalpha())
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:53]: try to use `filter`

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -41,6 +43,28 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            if list(word):
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:54]: Please, describe what do you want to check in this line?

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -104,6 +158,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+
+        if element not in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:164]: reverse the condition and do return as soon as posible

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -121,6 +180,18 @@ def decode(self, encoded_corpus: tuple[int, ...]) -> Optional[str]:
...
         """
+        if not isinstance(encoded_corpus, tuple):
+            return None
+
+        decoded_tokens = self._decode(encoded_corpus)
+        if not decoded_tokens:
+            return None
+
+        decoded_text = self._postprocess_decoded_text(decoded_tokens)
+        if not decoded_text:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:191]: Instead of this - return `decoded_text or None`

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -159,6 +246,21 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+
+        text_new = []
+
+        for i in decoded_corpus:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:254]: extend?

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -159,6 +246,21 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+            text_new.append(i)
+
+        if text_new[-1] == self._end_of_word_token:
+            text_new[-1] = '.'
+
+        text_string = ''.join(text_new)
+        text_string = text_string.replace(self._end_of_word_token, ' ')
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:261]: do it in return statement too

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -208,6 +314,18 @@ def build(self) -> int:
...
+        if not isinstance(self._encoded_corpus, tuple) or not self._encoded_corpus:
+            return 1
+
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+        if not isinstance(n_grams, tuple) or not n_grams:
+            return 1
+
+        for n_gram in set(n_grams):
+            n_grams_count = n_grams.count(n_gram)
+            context_count = len([context for context in n_grams if context[:-1] == n_gram[:-1]])
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:326]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -221,6 +339,15 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
 
         In case of corrupt input arguments, None is returned
         """
+        if not (isinstance(sequence, tuple) and sequence and len(sequence) >= self._n_gram_size - 1):
+            return None
+
+        next_tokens = {}
+        context = sequence[- self._n_gram_size + 1:]
+        for n_gram, freq in self._n_gram_frequencies.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:347]: very good

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
only for 4?

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:191]: fixed, I guess?

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:164]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:54]: I want to check whether every letter in a word is indeed a letter and not a number or punctuation mark.

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:254]: I didn't quite get what you mean

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py]: fixed?

File: lab_3_generate_by_ngrams/main.py (hunk 8 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py]: I'm not sure, but done

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py]: done?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -208,6 +306,19 @@ def build(self) -> int:
...
         """
+        if not isinstance(self._encoded_corpus, tuple) or not self._encoded_corpus:
+            return 1
+
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+        if not isinstance(n_grams, tuple) or not n_grams:
+            return 1
+
+        for n_gram in set(n_grams):
+            #context_count = len([context for context in n_grams if context[:-1] == n_gram[:-1]])
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:317]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -41,6 +43,30 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+
+        if not tokens:
+            return None
+
+        if text[-1].isalnum():
+            del tokens[-1]
+
+        return tuple(tokens)
+
+    # if word.isalpha():
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:66]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 12]
```diff
@@ -41,6 +43,30 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        for word in text.split():
+            word = list(filter(str.isalpha, word))
+            if word:
+                tokens.extend(word)
+                tokens.append(self._end_of_word_token)
+
+        if not tokens:
+            return None
+
+        if text[-1].isalnum():
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:61]: I understood my mistake. Fixed

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:250]: done

File: lab_3_generate_by_ngrams/main.py [hunk 13]
```diff
@@ -104,6 +162,10 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if isinstance(element, str) and len(element) == 1:
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:165]: fixed?

File: lab_3_generate_by_ngrams/main.py [hunk 14]
```diff
@@ -159,7 +242,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+
+        text_new = []
+
+        for i in decoded_corpus:
+            text_new.append(i)
+
+        if text_new[-1] == self._end_of_word_token:
+            del text_new[-1]
 
+        return ''.join(text_new).capitalize().replace(self._end_of_word_token, ' ') + '.'
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:256]: done?

File: lab_3_generate_by_ngrams/main.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:61]: I want to remove the separator after the last word if the text ends with a space or punctuation mark

File: lab_3_generate_by_ngrams/main.py [hunk 15]
```diff
@@ -76,6 +107,14 @@ def get_token(self, element_id: int) -> Optional[str]:
...
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        for key, value in self._storage.items():
+            if value == element_id:
+                return key
+
+        return None
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:117]: fixed?

File: lab_3_generate_by_ngrams/main.py [hunk 16]
```diff
@@ -159,7 +242,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+
+        text_new = []
+
+        for i in decoded_corpus:
+            text_new.append(i)
+
+        if text_new[-1] == self._end_of_word_token:
+            del text_new[-1]
```
[the remark this answers is not in the document]
Comment (kobyakovapolina) [reply lab_3_generate_by_ngrams/main.py:254]: Used it to remove the extra space

**Pull request closed without merging**